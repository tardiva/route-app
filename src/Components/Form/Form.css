.form__field {
    box-sizing: border-box;
    border: 1px solid #BEBEBE;
    padding: 7px;
    margin: 0;
    width: 100%;
    transition: all 0.3s ease-in-out;
    outline: none;
    font-size: 14px;
}

.form__field--error {
    border-color: #e83232;
}