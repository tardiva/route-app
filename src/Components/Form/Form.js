import React, { Component } from 'react';
import './Form.css';
import update from 'immutability-helper';

class CreateItemForm extends Component {

  constructor(props) {
    super(props);

    this.state = {
      itemName: '',
      touched: false
    };

    this.handleNameChange = this.handleNameChange.bind(this);
    // this.handleBlur = this.handleBlur.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  render() {
    const error = this.validate();
    return (
      <form onSubmit={this.onSubmit}>
        <input value={this.state.itemName}
               type="text"
               onChange={this.handleNameChange}
               className={"form__field " + (error && this.state.touched ? 'form__field--error' : '')}
               autoComplete="off"
               placeholder={'Новая точка маршрута'}/>
      </form>
    )
  }

  validate() {
    return this.state.itemName === '';
  }

  onSubmit(e) {
    e.preventDefault();

    if (this.validate()) {
      this.setTouched(true);
    } else {
      this.props.onSubmit(this.state.itemName );
      this.resetForm();
    }
  }

  handleNameChange(e) {
    const itemName = e.target.value;
    this.setState(() => ({ itemName }));
  };

  // handleBlur() {
  //   this.setTouched(true);
  // };

  setTouched(isTouched) {
    this.setState(update(this.state, {
      touched: {
        $set: isTouched
      }
    }));
  }

  resetForm() {
    this.setState(update(this.state, {
      itemName: {
        $set: ''
      },
      touched: {
        $set: false
      }
    }));
  }
}

export default CreateItemForm;