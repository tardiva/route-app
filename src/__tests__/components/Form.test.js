import React from 'react';
import { shallow } from 'enzyme';
import Form from '../../Components/Form/Form';

it('should render Form correctly', () => {
  const wrapper = shallow(<Form />);
  expect(wrapper).toMatchSnapshot();
});

it('should set item name on input change', () => {
  const value = 'New Valid Point';
  const wrapper = shallow(<Form />);
  const input = wrapper.find('input').at(0);
  input.simulate('change', {
    target: { value }
  });
  expect(wrapper.state('itemName')).toBe(value);
});

// it('should set touched on input blur', () => {
//   const wrapper = shallow(<Form />);
//   const input = wrapper.find('input').at(0);
//   input.simulate('blur');
//   expect(wrapper.state('touched')).toEqual(true);
// });

it('should call onSubmit prop for valid form submission', () => {
  const value = 'New Valid Point';
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<Form onSubmit={onSubmitSpy} />);
  wrapper.setState({itemName: value});
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(onSubmitSpy).toHaveBeenLastCalledWith(value);
});

it('should not call onSubmit prop for invalid form submission', () => {
  const value = '';
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<Form onSubmit={onSubmitSpy} />);

  wrapper.setState({itemName: value});
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  const input = wrapper.find('input').at(0);
  expect(input.hasClass('form__field--error')).toEqual(true);
  expect(onSubmitSpy).not.toHaveBeenCalled();
});