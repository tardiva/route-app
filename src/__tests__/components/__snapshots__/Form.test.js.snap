// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Form correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    className="form__field "
    onChange={[Function]}
    placeholder="Новая точка маршрута"
    type="text"
    value=""
  />
</form>
`;
